### PR TITLE
[DependencyInjection] Explain how to use the class itself as factory

### PR DESCRIPTION
**Pull request by OskarStark**
Follows
* https://github.com/symfony/symfony/pull/20943

### CODE REVIEW

File: service_container/factories.rst [hunk 1]
```diff
@@ -100,6 +100,81 @@ create its object:
...
+    // src/Email\NewsletterManager.php
+    namespace App\Email;
+
+    // ...
+
+    class NewsletterManager
+    {
+        private string $sender;
+
+        public static function create(string $sender): NewsletterManager
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
        public static function create(string $sender): self
```

File: service_container/factories.rst [hunk 2]
```diff
@@ -100,6 +100,81 @@ create its object:
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
 
+You can also use the ``factory`` option to create services from static methods.
+Lets suppose the ``NewsletterManager`` class has a ``create()`` method that needs
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: Let's
but I don't get your PR: the existing example (before this one) uses a static method already.

Comment (OskarStark) [reply service_container/factories.rst]: It uses a `NewsletterManagerFactory` class to create a `NewsletterManager` service. I create the `NewlsetterManager` service with the `create` method itself

Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
Let's suppose the ``NewsletterManager`` class has a ``create()`` method that needs
```

File: service_container/factories.rst [hunk 3]
```diff
@@ -100,6 +100,81 @@ create its object:
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
 
+You can also use the ``factory`` option to create services from static methods.
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: this is what the previous example does already

seeing the link you put in the description, I was expecting to read something about it - aka that the class can be omitted if the static factory method is on the same class than the created instance

Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
You can also use the ``factory`` option to create services from static methods
of the class itself.
```

File: service_container/factories.rst [hunk 4]
```diff
@@ -100,6 +100,82 @@ create its object:
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
 
+You can also use the ``factory`` option to create services from static methods
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: `When the static factory method is on the same class as the created instance, the class name can be omitted from the factory declaration.`

File: service_container/factories.rst [hunk 5]
```diff
@@ -100,6 +100,82 @@ create its object:
...
+
+    .. code-block:: yaml
+
+        # config/services.yaml
+        services:
+            # ...
+
+            App\Email\NewsletterManager:
+                # the first argument is the class and the second argument is the static method
+                # note that we omit the '@' prefix when using the class name as the first argument!
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
                # note that we omit the '@' prefix when using the class name as the first argument!
                # you can also remove the class name entirely and use 'null' instead
```

File: service_container/factories.rst [hunk 6]
```diff
@@ -100,6 +100,82 @@ create its object:
...
+            $newsletterManager = new self();
+            $newsletterManager->sender = $sender;
+            // ...
+
+            return $newsletterManager;
+        }
+    }
+
+You can use the ``factory`` option to define which method of which class must be
+called to create its object. In this case we don't need a dedicated factory class anymore:
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: You never need a dedicated classe to declare a static method so this sentence doesn't look super useful to me

File: service_container/factories.rst [hunk 7]
```diff
@@ -100,6 +100,82 @@ create its object:
...
+    .. code-block:: yaml
+
+        # config/services.yaml
+        services:
+            # ...
+
+            App\Email\NewsletterManager:
+                # the first argument is the class and the second argument is the static method
+                # note that we omit the '@' prefix when using the class name as the first argument!
+                factory: ['App\Email\NewsletterManager', 'create']
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: `factory: [null, 'create']`

File: service_container/factories.rst [hunk 8]
```diff
@@ -100,6 +100,82 @@ create its object:
...
+        <!-- config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <service id="App\Email\NewsletterManager">
+                    <!-- Note that we are using the "class" key here, not "service" -->
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: should be removed

File: service_container/factories.rst [hunk 9]
```diff
@@ -100,6 +100,83 @@ create its object:
...
+
+        # config/services.yaml
+        services:
+            # ...
+
+            App\Email\NewsletterManager:
+                # the first argument is the class and the second argument is the static method
+                # note that we omit the '@' prefix when using the class name as the first argument!
+                # you can also remove the class name entirely and use 'null' instead
+                factory: ['App\Email\NewsletterManager', 'create']
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
                factory: [null, 'create']
```

File: service_container/factories.rst [hunk 10]
```diff
@@ -100,6 +100,82 @@ create its object:
...
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <service id="App\Email\NewsletterManager">
+                    <!-- Note that we are using the "class" key here, not "service" -->
+                    <factory class="App\Email\NewsletterManager"
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: `<factory method="create"/>`

File: service_container/factories.rst [hunk 11]
```diff
@@ -100,6 +100,82 @@ create its object:
...
+        namespace Symfony\Component\DependencyInjection\Loader\Configurator;
+
+        use App\Email\NewsletterManager;
+
+        return function(ContainerConfigurator $containerConfigurator) {
+            $services = $containerConfigurator->services();
+
+            // Note that we are not using service()
+            $services->set(NewsletterManager::class)
+                ->factory([NewsletterManager::class, 'create']);
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: `->factory([null, 'create']);`

File: service_container/factories.rst [hunk 12]
```diff
@@ -100,6 +100,80 @@ create its object:
...
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <service id="App\Email\NewsletterManager">
+                    <!-- Note that we are using the "class" key here, not "service" -->
+                    <factory class="App\Email\NewsletterManager"
+                        method="create"
+                    />
+                </service>
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
                    <factory method="create"/>
                </service>
```

File: service_container/factories.rst [hunk 13]
```diff
@@ -100,6 +100,77 @@ create its object:
...
+        namespace Symfony\Component\DependencyInjection\Loader\Configurator;
+
+        use App\Email\NewsletterManager;
+
+        return function(ContainerConfigurator $containerConfigurator) {
+            $services = $containerConfigurator->services();
+
+            // Note that we are not using service()
+            $services->set(NewsletterManager::class)
+                ->factory([NewsletterManager::class, 'create']);
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
                ->factory([null, 'create']);
```

File: service_container/factories.rst [hunk 14]
```diff
@@ -100,6 +100,77 @@ create its object:
...
+            $newsletterManager = new self();
+            $newsletterManager->sender = $sender;
+            // ...
+
+            return $newsletterManager;
+        }
+    }
+
+You can use the ``factory`` option to define which method of which class must be
+called to create its object. In this case we don't need a dedicated factory class anymore:
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
called to create its object:
```

File: service_container/factories.rst [hunk 15]
```diff
@@ -100,6 +100,77 @@ create its object:
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
 
+You can also use the ``factory`` option to create services from static methods
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
When the static factory method is on the same class as the created instance, the class name can be omitted from the factory declaration.
```

File: service_container/factories.rst [hunk 16]
```diff
@@ -100,6 +100,77 @@ create its object:
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
```
Comment (OskarStark) [inline service_container/factories.rst:102]:
```suggestion
Using the Class as Factory Itself
---------------------------------

```

File: service_container/factories.rst [hunk 17]
```diff
@@ -99,6 +99,79 @@ create its object:
     only depends on the object that is returned by the factory. However,
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
+Using the Class as Factory Itself
+---------------------------------
+
+When the static factory method is on the same class as the created instance, the class name can be omitted from the factory declaration.
+of the class itself.
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: drop

File: service_container/factories.rst [hunk 18]
```diff
@@ -99,6 +99,79 @@ create its object:
...
+Let's suppose the ``NewsletterManager`` class has a ``create()`` method that needs
+to be called to create the object and need a send email address::
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: a sender

File: service_container/factories.rst (hunk 17 above)
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
```

File: service_container/factories.rst [hunk 19]
```diff
@@ -99,6 +99,78 @@ create its object:
     only depends on the object that is returned by the factory. However,
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
+Using the Class as Factory Itself
+---------------------------------
+
+When the static factory method is on the same class as the created instance, the class name can be omitted from the factory declaration.
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
When the static factory method is on the same class as the created instance,
the class name can be omitted from the factory declaration.
```

File: service_container/factories.rst [hunk 20]
```diff
@@ -99,6 +99,79 @@ create its object:
...
+        {
+            $newsletterManager = new self();
+            $newsletterManager->sender = $sender;
+            // ...
+
+            return $newsletterManager;
+        }
+    }
+
+You can use the ``factory`` option to define which method of which class must be
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: You can omit the class on the factory declaration:

File: service_container/factories.rst [hunk 21]
```diff
@@ -99,6 +99,79 @@ create its object:
     only depends on the object that is returned by the factory. However,
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
+Using the Class as Factory Itself
+---------------------------------
+
+When the static factory method is on the same class as the created instance,
+the class name can be omitted from the factory declaration.
+Let's suppose the ``NewsletterManager`` class has a ``create()`` method that needs
+to be called to create the object and need a send email address::
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
to be called to create the object and need a sender::
```

File: service_container/factories.rst [hunk 22]
```diff
@@ -99,6 +99,79 @@ create its object:
     only depends on the object that is returned by the factory. However,
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
+Using the Class as Factory Itself
```
Comment (nicolas-grekas) [inline service_container/factories.rst:103]: missing blank line

Comment (OskarStark) [inline service_container/factories.rst:103]:
```suggestion

Using the Class as Factory Itself
```

File: service_container/factories.rst [hunk 23]
```diff
@@ -100,6 +100,80 @@ create its object:
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
 
+Using the Class as Factory Itself
+---------------------------------
+
+When the static factory method is on the same class as the created instance,
+the class name can be omitted from the factory declaration.
+Let's suppose the ``NewsletterManager`` class has a ``create()`` method that needs
+to be called to create the object and need a sender::
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: needs

File: service_container/factories.rst [hunk 24]
```diff
@@ -100,6 +100,80 @@ create its object:
...
+        {
+            $newsletterManager = new self();
+            $newsletterManager->sender = $sender;
+            // ...
+
+            return $newsletterManager;
+        }
+    }
+
+You can use the ``factory`` option to define which method of which class must be
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
You can omit the class on the factory declaration:
```

File: service_container/factories.rst [hunk 25]
```diff
@@ -100,6 +100,80 @@ create its object:
...
+called to create its object:
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
```

File: service_container/factories.rst [hunk 26]
```diff
@@ -100,6 +100,80 @@ create its object:
...
 
+Using the Class as Factory Itself
+---------------------------------
+
+When the static factory method is on the same class as the created instance,
+the class name can be omitted from the factory declaration.
+Let's suppose the ``NewsletterManager`` class has a ``create()`` method that needs
+to be called to create the object and need a sender::
+
+    // src/Email\NewsletterManager.php
```
Comment (nicolas-grekas) [inline service_container/factories.rst]: forward slash before News

Comment (OskarStark) [reply service_container/factories.rst]:
```suggestion
    // src/Email/NewsletterManager.php
```

File: service_container/factories.rst (hunk 23 above)
Comment (OskarStark) [reply service_container/factories.rst]: really?

Comment (nicolas-grekas) [reply service_container/factories.rst]: yes - the subject is the method, no?

File: service_container/factories.rst [hunk 27]
```diff
@@ -100,6 +100,79 @@ create its object:
     the configured class name may be used by compiler passes and therefore
     should be set to a sensible value.
 
+Using the Class as Factory Itself
+---------------------------------
+
+When the static factory method is on the same class as the created instance,
+the class name can be omitted from the factory declaration.
+Let's suppose the ``NewsletterManager`` class has a ``create()`` method that needs
+to be called to create the object and need a sender::
```
Comment (OskarStark) [inline service_container/factories.rst]:
```suggestion
to be called to create the object and needs a sender::
```

File: service_container/factories.rst (hunk 23 above)
Comment (OskarStark) [reply service_container/factories.rst]: You are right, but needs twice sounds a bit weird maybe

**Review by nicolas-grekas — APPROVED** [review]

**Comment by OskarStark** [comment]
Thanks @nicolas-grekas 👍

File: service_container/factories.rst (hunk 23 above)
Comment (nicolas-grekas) [reply service_container/factories.rst]: accepts, expects, etc

**Comment by nicolas-grekas** [comment]
On L69, we mention this capability. I suppose we should remove this previous mention.

**Comment by OskarStark** [comment]
Done

**Pull request merged**